### ISSUE TITLE
Can't log in, if using "language in URL" option

### ISSUE THREAD

**Issue by herrvigg** [herrvigg/qtranslate-x#141]
<a href="https://github.com/taai"><img src="https://avatars1.githubusercontent.com/u/90734?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [taai](https://github.com/taai)**
_Thursday Apr 23, 2015 at 09:47 GMT_
_Originally opened as https://github.com/qTranslate-Team/qtranslate-x/issues/141_

----

I store the WordPress in `/wp` directory, so it's easy to upgrade it with Composer. So, the log-in address should be `/wp/wp-login.php`, but when `Hide URL language information for default language.` checkbox is **not** checked, it redirects to non-existing `/en/wp/wp-login.php`.

There probably is a bug also when hiding the default language in URL - I have noticed that sometimes it is not possible to switch language for some reason. That's why I wanted to always include language code in URL.

**Comment by herrvigg** [comment]
<a href="https://github.com/Grafcom"><img src="https://avatars0.githubusercontent.com/u/11356362?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Grafcom](https://github.com/Grafcom)**
_Thursday Apr 23, 2015 at 14:31 GMT_

----

@taai,

see this support ticket:
https://wordpress.org/support/topic/qtranslate-x-navigation-menu-not-working-properly?replies=5

**Comment by herrvigg** [comment]
<a href="https://github.com/taai"><img src="https://avatars1.githubusercontent.com/u/90734?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [taai](https://github.com/taai)**
_Thursday Apr 23, 2015 at 17:41 GMT_

----

@Grafcom I don't think it's the same. My website is meant to work in the root directory, just the WordPress files are being stored in `/wp/` directory. Everything works well when `Hide URL language information for default language` option is turned on.
These are the constants that are being set in `/wp-config.php`:

``` php
define('WP_CONTENT_DIR', __DIR__ . '/wp-content');
define('WP_CONTENT_URL', 'http://' . $_SERVER['SERVER_NAME'] . '/wp-content');
define('WP_SITEURL', 'http://' . $_SERVER['SERVER_NAME'] . '/wp');
define('WP_HOME', 'http://' . $_SERVER['SERVER_NAME']);
```

I think that real PHP files like `/wp/wp-login.php` should not be considered as a translatable URLs. So, there should be a `is_file($path) return;` check built in this plugin!

**Comment by herrvigg** [comment]
<a href="https://github.com/Grafcom"><img src="https://avatars0.githubusercontent.com/u/11356362?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Grafcom](https://github.com/Grafcom)**
_Thursday Apr 23, 2015 at 17:45 GMT_

----

@taai,
- My website is meant to work in the root directory, just the WordPress files are being stored in /wp/ directory.

That's exactly where that topic is about. See the .htaccess en index.php settings

**Comment by herrvigg** [comment]
<a href="https://github.com/Grafcom"><img src="https://avatars0.githubusercontent.com/u/11356362?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Grafcom](https://github.com/Grafcom)**
_Thursday Apr 23, 2015 at 17:57 GMT_

----

@taai,
only to login in this case you must use /wp-admin and not /wp-login

**Comment by herrvigg** [comment]
<a href="https://github.com/taai"><img src="https://avatars1.githubusercontent.com/u/90734?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [taai](https://github.com/taai)**
_Friday Apr 24, 2015 at 08:54 GMT_

----

@Grafcom I'm sorry, it didn't work for me and I don't see how is this supposed to work at all.

The `/wp/wp-login.php` **is** being executed, because it's a **real PHP file**. I don't see how it is Apache's configuration fault. The problem, in this case, is that _qtranslate-x_ thinks that it should redirect to URL with language code in it - to non-existing URL `/en/wp/wp-login.php` which then will be passed to `/index.php`. And that, in my opinion, is wrong!

I suggest to check if the URL that _qtranslate-x_ is about to redirect from should be checked if it is a real file or not and, if it is a real file, not to redirect. Have look at it right there:
https://github.com/qTranslate-Team/qtranslate-x/blob/master/qtranslate_core.php#L94

I probably could contribute and code the solution, just I don't know all the internals of this plugin and if my solution will not break something I'm not aware of. Should I do it?

Here is the dump of `$url_info` variable just before `wp_redirect($target);`:

```
array(12) {
  ["pagenow"]=>
  string(12) "wp-login.php"
  ["scheme"]=>
  string(4) "http"
  ["host"]=>
  string(11) "example.com"
  ["path"]=>
  string(16) "/wp/wp-login.php"
  ["original_url"]=>
  string(16) "/wp/wp-login.php"
  ["path-base"]=>
  string(0) ""
  ["path-base-length"]=>
  int(0)
  ["doing_front_end"]=>
  bool(true)
  ["wp-path"]=>
  string(16) "/wp/wp-login.php"
  ["doredirect"]=>
  string(33) "language needs to be shown in url"
  ["language"]=>
  string(2) "en"
  ["set_cookie"]=>
  bool(true)
}
```

If I'm going to contribute on this, which value should I use to check if it is a real file - the `$url_info["path"]` or `$url_info["wp-path"]` ?

**Comment by herrvigg** [comment]
<a href="https://github.com/Grafcom"><img src="https://avatars0.githubusercontent.com/u/11356362?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Grafcom](https://github.com/Grafcom)**
_Sunday Apr 26, 2015 at 14:06 GMT_

----

@taai,
Do I understand correctly?
the website works as you want?
WordPress stored in /wp directory. Site accessible from the root.

The only problem, when Hide URL language information for default language checkbox is not checked, yoursite.com/wp/login or yoursite.com/wp/wp-login.php does not work. Still correct?

But how it is with yoursite.com/wp/wp-admin that does work?

**Comment by herrvigg** [comment]
<a href="https://github.com/taai"><img src="https://avatars1.githubusercontent.com/u/90734?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [taai](https://github.com/taai)**
_Monday Apr 27, 2015 at 15:31 GMT_

----

@Grafcom Yes, you have understood the problem correctly. No, even "yoursite.com/wp/wp-admin" doesn't work. Except if you are already logged in.

**Comment by herrvigg** [comment]
<a href="https://github.com/Grafcom"><img src="https://avatars0.githubusercontent.com/u/11356362?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Grafcom](https://github.com/Grafcom)**
_Monday Apr 27, 2015 at 19:13 GMT_

----

@taai,
- I have noticed that sometimes it is not possible to switch language for some reason. That's why I wanted to always include language code in URL.

This is actually the first problem to be studied/solved. Before it makes sense to be busy with that login issue.
Would you open a ticket with all the details of that problem on this support forum?
https://wordpress.org/support/plugin/qtranslate-x
Thanks

**Comment by herrvigg** [comment]
<a href="https://github.com/taai"><img src="https://avatars1.githubusercontent.com/u/90734?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [taai](https://github.com/taai)**
_Tuesday Apr 28, 2015 at 07:00 GMT_

----

@Grafcom A ticket is now open at https://wordpress.org/support/topic/cant-log-in-if-using-language-in-url-option?replies=1

But why wouldn't it be sensible not to redirect when the URL (like `/wp/wp-login.php`) is a real PHP file? It's not like `/de/wp/wp-login.php` is working anyway...

**Comment by herrvigg** [comment]
<a href="https://github.com/taai"><img src="https://avatars1.githubusercontent.com/u/90734?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [taai](https://github.com/taai)**
_Tuesday Apr 28, 2015 at 11:44 GMT_

----

@Grafcom I did manage to repeat the weird bug when being unable to switch back to the main language.

Go to `http://prepage.eu/at/recommendations` and then `http://prepage.eu/recommendations` - it stays at `http://prepage.eu/at/recommendations`
I'm using Chrome in incognito mode.

Dump of the `$url_info` right before `wp_redirect($target)`:

```
array(15) {
  ["pagenow"]=>
  string(9) "index.php"
  ["scheme"]=>
  string(4) "http"
  ["host"]=>
  string(10) "prepage.eu"
  ["path"]=>
  string(16) "/recommendations"
  ["original_url"]=>
  string(16) "/recommendations"
  ["path-base"]=>
  string(0) ""
  ["path-base-length"]=>
  int(0)
  ["doing_front_end"]=>
  bool(true)
  ["wp-path"]=>
  string(16) "/recommendations"
  ["http_referer"]=>
  string(36) "http://prepage.eu/at/recommendations"
  ["internal_referer"]=>
  bool(true)
  ["referer_language"]=>
  string(2) "at"
  ["language"]=>
  string(2) "at"
  ["set_cookie"]=>
  bool(true)
  ["doredirect"]=>
  string(22) "$url_orig != $url_lang"
}
```

**Comment by herrvigg** [comment]
<a href="https://github.com/Grafcom"><img src="https://avatars0.githubusercontent.com/u/11356362?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Grafcom](https://github.com/Grafcom)**
_Tuesday Apr 28, 2015 at 13:01 GMT_

----

@taai,
See here for my reaction:
https://wordpress.org/support/topic/cant-log-in-if-using-language-in-url-option?replies=3

**Comment by herrvigg** [comment]
<a href="https://github.com/taai"><img src="https://avatars1.githubusercontent.com/u/90734?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [taai](https://github.com/taai)**
_Tuesday Apr 28, 2015 at 14:32 GMT_

----

@Grafcom I have posted the solution for the problem with switching to default language here: https://wordpress.org/support/topic/cant-log-in-if-using-language-in-url-option?replies=4#post-6882153

Replace this: https://github.com/qTranslate-Team/qtranslate-x/blob/723f34c002b4023c93b8af84b9263665af8b4d12/qtranslate_core.php#L238-L248
with this:

``` php
if( !empty($url_info['wp-path'])) {
    if(preg_match('!^/([a-z]{2})(/|$)!i',$url_info['wp-path'],$match)) {
        $lang = qtranxf_resolveLangCase($match[1],$doredirect);
        if($lang){
            $url_info['lang_url'] = $lang;
            $url_info['wp-path'] = substr($url_info['wp-path'],3);
            $url_info['doing_front_end'] = true;
            if(WP_DEBUG){
                $url_info['url_mode'] = 'pre-path';
            }
        }
    }
    elseif ($q_config['hide_default_language'] && !empty($q_config['default_language']))
    {
        $lang = $q_config['default_language'];
        $url_info['lang_url'] = $lang;
        $url_info['doing_front_end'] = true;
        if(WP_DEBUG){
            $url_info['url_mode'] = 'pre-path';
        }
    }
}
```

That kinda resolves my problem, if I hide the default language in URL. But if not hiding the default language? So, can we move on to my initial problem? :wink:

**Comment by herrvigg** [comment]
<a href="https://github.com/taai"><img src="https://avatars1.githubusercontent.com/u/90734?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [taai](https://github.com/taai)**
_Tuesday Apr 28, 2015 at 16:04 GMT_

----

@Grafcom I think I found a solution.

Have a look at this code right here:
https://github.com/qTranslate-Team/qtranslate-x/blob/723f34c002b4023c93b8af84b9263665af8b4d12/qtranslate_core.php#L869-L873

This is my solition:

``` php
$match_path = $path;

if (defined('WP_SITEURL'))
{
    $wp_siteurl_path = parse_url(WP_SITEURL, PHP_URL_PATH);
    $len = strlen($wp_siteurl_path);

    if (substr($path, 0, $len + 1) === $wp_siteurl_path.'/')
    {
        $match_path = substr($path, $len);
    }
}

if(preg_match('#^/(wp-.*\.php|wp-admin/|xmlrpc.php|.*sitemap.*|robots.txt|oauth/)#', $match_path)){
    $language_neutral_path_cache[$path] = true;
    //qtranxf_dbg_log('qtranxf_language_neutral_path: preg_match: path='.$path);
    return true;
}
```

Two issues resolved, huh? :wink:

**Comment by herrvigg** [comment]
<a href="https://github.com/johnclause"><img src="https://avatars2.githubusercontent.com/u/5491361?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [johnclause](https://github.com/johnclause)**
_Tuesday Apr 28, 2015 at 21:55 GMT_

----

Hi @taai: are you dong your tests in the latest version from GitHub? The reason I ask is that the latest version does not redirect to `/en/wp/wp-login.php` as you mentioned in the first report, and I could not reproduce your problem so far under the latest version from [GitHub](https://github.com/qTranslate-Team/qtranslate-x): https://github.com/qTranslate-Team/qtranslate-x/archive/master.zip.

Here is the sequence I did:

In /wp-admin/options-general.php, I have:
-  WordPress Address (URL): http://example.com/wp
-  Site Address (URL): http://example.com

I hope I understood the description of your setup correctly, did I?

Then I do:
- logout
- /wp/wp-login.php - shows  login page in the currently chosen front language, and login action works as it is supposed.
- /wp/wp-login.php?lang=xx - switches currently chosen front language to 'xx', redirects to /wp/wp-login.php, shows login page in the new language, and login action works as it is supposed.
- /wp/xx/wp-login.php - does same action as above, since I have 'Use Pre-Path Mode' on configuration page /wp-admin/options-general.php?page=qtranslate-x.
- /xx/wp/wp-login.php - ends up with a front-end 404 error, as it is supposed to.

So, what exactly does not work for you? Please, describe the steps again in a clean way, and remember to take the latest version. Please, remove qtx that you have now and download the fresh zip file https://github.com/qTranslate-Team/qtranslate-x/archive/master.zip, even if you have done it before - this way you will be sure that there is no accidental usage of a previous version.

**Comment by herrvigg** [comment]
<a href="https://github.com/taai"><img src="https://avatars1.githubusercontent.com/u/90734?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [taai](https://github.com/taai)**
_Thursday Apr 30, 2015 at 10:53 GMT_

----

@johnclause No, I was using version 3.2.9 downloaded automatically from http://wpackagist.org via Composer.
Don't know how, but it looks like the problem **is solved** in latest version 3.2.9.9.7 which is downloadable from GitHub. I think I will load this plugin straight from GitHub instead of http://wpackagist.org . You should release updates to _Wordpress Plugin Directory_ more often. :wink:
